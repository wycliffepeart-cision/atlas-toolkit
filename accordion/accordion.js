import { _ as __decorate } from '../tslib.es6-806de115.js';
import { Template } from '../decorators/template.js';
import AccordionTemplate from './template.js';

let Accordion = class Accordion extends HTMLElement {
};
Accordion = __decorate([
    Template(AccordionTemplate)
], Accordion);

export { Accordion };
