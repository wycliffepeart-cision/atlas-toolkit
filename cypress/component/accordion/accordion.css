#demo-according-body {
  padding: 10rem;
  text-align: center;
  border-radius: 8px;
  background-color: #fff;
  border: 1px solid #ddd;
}
