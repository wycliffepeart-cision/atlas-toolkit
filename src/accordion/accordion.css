atlas-accordion {
    display: block;
    border-radius: 6px;
    border: 1px solid #ddd;
    transform: translateZ(0);
    background-color: #f8f8f8;
}

atlas-accordions {
    transition: 0.4s all;
}

atlas-accordion button {
    margin: 0;
    outline: 0;
    border: 0;
    width: 100%;
    display: grid;
    outline: none;
    cursor: pointer;
    font-size: 18px;
    font-weight: 700;
    line-height: 24px;
    padding: 18px 20px;
    position: relative;
    text-align: left;
    background-color: transparent;
    grid-template-columns: 1fr auto;
}

atlas-accordion-content {
    padding: 0 20px 20px 20px;
    display: block;
}

atlas-accordion::part(title) {
    text-align: left;
    display: block;
}

atlas-accordion::part(content) {
    padding: 0 20px 20px 20px;
    display: block;
}

atlas-accordions {
    display: flex;
    flex-direction: column;
    gap: 1rem;
}

atlas-accordions atlas-accordion[collapse='showing'],
atlas-accordions atlas-accordion[collapse='show'] {
    display: block;
    transition: 0.4s margin;
}

atlas-accordions atlas-accordion[collapse='showing']:first-child,
atlas-accordions atlas-accordion[collapse='show']:first-child {
    margin-top: 0;
}

atlas-accordions atlas-accordion[collapse='showing']:first-child,
atlas-accordions atlas-accordion[collapse='show']:last-child {
    margin-bottom: 0;
}
