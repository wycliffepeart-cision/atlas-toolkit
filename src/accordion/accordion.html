<template>
	<slot part="header" name="header"></slot>
	<slot part="body"></slot>
</template>
