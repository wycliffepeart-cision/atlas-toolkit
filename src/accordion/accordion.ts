import template from './template.html';
import { makeId } from '../utils/utils';

export class Accordion extends HTMLElement {
	private readonly uniqueId;

	private readonly targetSelector;

	private callback;

	constructor() {
		super();

		this.uniqueId = makeId(10);
		this.id = this.uniqueId;

		this.targetSelector = (collapse) => `${this.tagName}[id="${this.uniqueId}"][collapse="${collapse}"]::part(collapsable)`;

		const templateElement = new DOMParser().parseFromString(template, 'text/html').querySelector('template');

		if (templateElement === undefined || templateElement === null) {
			throw new Error('Template undefined');
		}

		const { content } = templateElement;
		const shadowRoot = this.attachShadow({ mode: 'open' });
		shadowRoot.replaceChildren(content.cloneNode(true));
	}

	resetStyle() {
		this.setAttribute('collapse', 'showing');
		setTimeout(() => {
			this.setAttribute('collapse', 'hidden');
		}, 10);
	}

	connectedCallback() {
		const target = this.shadowRoot.querySelector('[part="collapsable"]');
		const style = document.createElement('style');
		style.id = this.uniqueId;

		const cssFn = () =>
			(style.textContent = `
      atlas-accordion::part(collapsable){
        transition: .5s all;
      }
      ${this.targetSelector('hidden')} { height: ${0}px; overflow: hidden;  }
      ${this.targetSelector('showing')} { height: ${target.scrollHeight}px; overflow: hidden }
      ${this.targetSelector('show')} { height: auto; overflow: hidden }
    `);

		document.head.append(style);

		cssFn();

		this.setAttribute('collapse', !this.getAttribute('collapse') || this.getAttribute('collapse') === 'hidden' ? 'show' : 'hidden');

		if (!this.callback) {
			this.callback = () => {
				setTimeout(() => {
					if (this.getAttribute('collapse') === 'showing') this.setAttribute('collapse', 'show');
				}, 10);
			};
		}

		target.addEventListener('transitionend', this.callback.bind(this));
		const btn = this.shadowRoot.querySelector('button');

		this.shadowRoot.querySelector('button').addEventListener('click', (e) => {
			if (!this.getAttribute('collapse') || this.getAttribute('collapse') === 'hidden') {
				btn.setAttribute('collapse', 'show');
				this.setAttribute('collapse', 'showing');
			} else if (this.getAttribute('collapse') === 'show') {
				btn.setAttribute('collapse', 'hidden');
				this.resetStyle();
			}
		});
	}
}
