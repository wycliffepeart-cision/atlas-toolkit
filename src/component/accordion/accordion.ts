import { Template } from '../../decorators/template';
import AccordionTemplate from './template.html';

@Template(AccordionTemplate)
export class Accordion extends HTMLElement {}

if (!customElements.get('atlas-accordion')) customElements.define('atlas-accordion', Accordion);