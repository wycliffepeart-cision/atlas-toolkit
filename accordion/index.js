import { Accordion } from './accordion.js';
import '../tslib.es6-806de115.js';
import '../decorators/template.js';
import './template.js';

if (!customElements.get('atlas-accordion'))
    customElements.define('atlas-accordion', Accordion);
