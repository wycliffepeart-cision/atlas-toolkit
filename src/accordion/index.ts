import './content'
import './collapsable'
import './accordions'
import { Accordion } from './accordion';
import './accordion.css'

if (!customElements.get('atlas-accordion')) customElements.define('atlas-accordion', Accordion);

