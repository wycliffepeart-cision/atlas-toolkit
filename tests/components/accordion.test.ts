import '../../src/accordion';

const getShadowDomContent = (element) => element.shadowRoot.innerHTML;

describe('Accordion Component', () => {
  const ACCORDION_TAG = 'atlas-accordion';

  it('is correctly appended in the DOM', async () => {
    await customElements.whenDefined(ACCORDION_TAG);
    const snackbarElement = document.createElement(ACCORDION_TAG);
    document.body.appendChild(snackbarElement);

    expect(snackbarElement).toBeInTheDocument();
  });

  it('is capable of being selected', async () => {
    await customElements.whenDefined(ACCORDION_TAG);
    const snackbarElement = document.createElement(ACCORDION_TAG);
    document.body.appendChild(snackbarElement);

    const selectedElement = document.querySelector(ACCORDION_TAG);
    (expect(selectedElement) as any).toBeInTheDocument();
  });

  it('has the correct content when created', async () => {
    await customElements.whenDefined(ACCORDION_TAG);
    const snackbarElement = document.createElement(ACCORDION_TAG);
    document.body.appendChild(snackbarElement);

    (expect(snackbarElement) as any).toContainHTML('<atlas-accordion></atlas-accordion>');
    (expect(snackbarElement) as any).toBeEmptyDOMElement();
    (expect(getShadowDomContent(snackbarElement)) as any).toMatchSnapshot();
  });
});
