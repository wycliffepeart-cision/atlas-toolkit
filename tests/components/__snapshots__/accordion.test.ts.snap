// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion Component has the correct content when created 1`] = `
<slot part="header"
      name="header"
>
</slot>
<slot part="body">
</slot>
`;
